### ISSUE TITLE
Bug: [no-unused-vars] still false positive if name is type/value merged and exported

### ISSUE THREAD

**Issue by Dannymx** [typescript-eslint/typescript-eslint#7336]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I know that previously [this bug issue](https://github.com/typescript-eslint/typescript-eslint/issues/6188) was submitted related to false positives with types merged and exported, today I still see the same error in my own repo.

I've tried:
- Updating all packages 
- Removing packages that could have interfered, so to narrow the issue to `typescript-eslint` only.
- Tried to create a basic snippet where I'm seeing the error locally (vscode) onto TS Playground, but I don't see the error there, here's the link [Playground](https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=C4TwDgpgBAQghgZwJYGMDyAjAVhFwoC8UA3gFBRTAQLABcUNATkgHYDmA3KQL5ekQAPMAHtG%2BUJFiJUAFXDQiAJVyiAJgB5yDYM3YAaLQApiUFnAC2EPQwA2AVzZRu9E2cv0mrTrYcedXpwBKQgA%2BKWR0bFxgUhCOIA&eslintrc=N4KABGBED2BuCmAnRBLAJvAzpAXGA2uBGAPQlgDC0AdgGYoDmArogIYAuKNYt0iYAFQCeAB3gBlAMaoR7HigA2WIhFDFikekux5C69ZABUhkoYB07bABoV%2BoyfOWAHpFsQAujf1QRCpgxRqHQI3DQABdlEsaRRZAFosBUD2SC9vUnJIJmomTHg0OJQAWxE%2BS1TQiEh2VkUAd0C0SUxrSqhMYt94QpKyuMwy129PUMh4J3Z4ajRgvXTIX39AnBr6xubMEkR4SWgioqmMNAr0qFrEACNqC7jIsUwY2RP56nH2El3tuLr4G9gUGoKVqnBbbdicJBDfQjbwLViIPKIADysi4QVwYDU8xEiGgACsdik8JAzCRLLs6IwzHiBtQoeoAL5pAyIJjaDFY2HbViSd4YTDsVm8liBBhxVgtRjUA7UIlQaC0WipDJgABqD2gGDAaGgWGoAHI5JgmCJSog5KwmOw9hwUJJWAoFEJtVhBUxhahqAwrGAAfrMGBWGARLVqGBrYG0GhA2BXnUwLB4ShWBclG1IDSnOKAIzZoQkVjUSQACz4hUwcUTSWOxIVSp9ZDAADk3tSA7l4GBiygA9A6mHkkhqA6wElqABrMCYIQC%2BBFdPc3lbeAARyYKC%2BGFoloU7DiOOgImCMEVysbAEFHX3YzQ4lvAvkwIu5Aej4H29N4PRXmgF-Aee8mb7riR5xNQ0D9Di-5oKKGInvWKoAPqsKaFiYE4YAdgGmBQaw0aHkgHB8IG0yBkC0A%2Bs%2BcSltA45xLwiDziCVE0eOmzjMWloCigCB3vAR5wXWZ7kAAkkWfDbLyTpPvxZSxvA%2BSPhgYifkWzoNOwxbNq2NJPtAVpIOmYSvBMJAme84E9GK8BKDKcrwcJYAAOqdpILDbLK0ngXIuSivJpnFKwDCdhcOxcZ2AK%2BpgBpGiaZqTNGGlaeI4gAOJgEUmrwEZdzRDIe6JMkHx7EUrB3oWDBKIJp4NuQ56wNA6BgBStBJLyT5siF8DsD8UxgAAopgY4WqRAAKYIQogOVRA8%2BUJENRUUh0s6yrcUQ9PFx5ILiiCOf1QQsJ2AAGnRyblR1RZheSJcWfV%2BgGrySFgmDwkI6aneaJBQbQSB8dubIFU48XVQhjZNqwBzRuM8WXf%2BHRIOG0AyduvLEZa1qlZw9qOm9IKrAoDTTBsHwKBK0UQ1gcR8Bgu26G0VR1PCdLMuknKnPKip2smCjiGUsF4G62Xs2ADJtJ4KoAGLQFe8aaZ2L0HGA1NIH5ErhrdSuc5I3PBn4ARhkdUHgigv2LPrty1AT6wtEd6YdCUSgbX0AyfR95TEttfCOSJvTms1ND0MwbCcNwDFgEdhWrWbgT9J0jtu-0ZS2yC9tdE75qJ590NlFtyBe7VA1A3JLWMCwtqh8REcLVHesx6n8e%2B3uLvsMn8wRDNjwFdXFkQdkHYFImCJwZ7u0s%2BojZZDk13p5YZm91P%2BQz7nO3PN4E999PbubJZG%2BL4PsxtI2VQj6v%2BhH5iZyIAwmAiQw4HbKNHCTIgdLEgAeohkAi4f5AwvMz4kFoNkXkaI4i7F6K8Va95qAAjRCDU%2BBg3YkGVrTDmSpQii0ZCzNmVQtBYAxHMWE5gaQII0MQ9CpCqiOGBPMahLhQh-xZJ1YIOCND126AnZuJAt4EPplAE%2BY99CsNhAwXEJoD7C3UIQyREBGyP0kOOIKWAzDh2fBdbYpMErBh5Io4KAZwGdnoAidgZg%2BGwjfs%2BSh7NIBvzCAAfgADoOLqFY04jYxLP2HAobRCilGYFMTIgwb8AAU4DSiQMsAAH1YpgSJmBOLbDQJEq0ihYkUkmBMWJt0FBiARHE602wACUwSSBmEMJEgAJIU1x6RGziHQJ2L8P12pbwCYEqob8nFMAAAy9O6TUte5BH4eTkK0sAo0rThzMGYC6pMBRtPaTYpxZhlnBLsQAQhIHYqpAy7CdIcSsg5WyKm7PHuQJEct%2BAaNtAgX0jd-HjMmQrbovAFA0zuZtEi0YjozNHBKExZi9nLJIGsgAvGUkgxSNnHOqYI4WSyDlrM2ds2FgKgnAu2ac4gdTIhKA%2BTnBZgSbFmAANTLMwHYjYJy0VgHcICulwtMHpAZfoJlEAmS2EbFQSkQdy5hjDpMbiXpbCsM0IofBdMQTGFMBYV0aErH2BlYKkxzh6TEEYRoaO6JJXzAJAKBVeq9wMUxpwL0CrlWijiEkC4bBEBCANa6O8ew1UeEEWMCYhwJHYlrtQHAhrlzgJlEcBVWq-WOuNU-UUAaSqHHyCGn1KxXSWutba-Mli4U%2BATYap1RRo37FjccBhtgmV0oZEAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkqyARhWANrg20AmCkANOx0jd4AOmR4sAQwYY%2BAmpHQki6AB55I8gLr9B8WJwCusyouVqNuhfqMYASktipDBYs0hED6OYMyPUAG7oAFIkxACyBsbelHiYhuhWtH76QQDCCPDoRHhksfGJ8pCIfiRY6cQkeACiRIbweWBxCUlQVZjI0Br5LUWSqADukgCeJADKcZ3dTQWtkO1TAHLGqGkAFujQANaNzYWCngCqRNAb2%2BicAJJEnGoXAILQ0Eq7s0Wel0j00OL3RMPMPZzD5fTriBx4QyYUiAt4HWCfRDfcQAFTWyFevXhx0MZU4ABlYNB%2Bpj9gojnU8QAFSSYSRZCSYUlzfqoWADY5%2BSSnaQYDK3ZgAMxJZNongAYqy8GtMLBDABzNZpSRlEjXMYDcSnWFYhR4Wny9DTMwAVh8CjleAAIshMO5LeauJtULTJK4iDrRW05ZhnuFJIhPXMlJFbOhrozYIGel7IKz2WNhjkNgRoFb0MLjHhEbBMLkg31UGyBqEhSK5iQtshEPjGOtNlsC4JBbnnhlSBiJDllSRkER5ddxWh0It6UomwoAFYkVTuEpKcreYGwarwcQT2gYpxui6h6LMop96ApbL61AbqAMFXoQ6Yc%2BmESOqCIN1rRrUDhQAACAHoAFTMGwn4KCI-6aMBOjyLQv6IIYMidGBlBAcBUCgbB8HQGBUFgJBwEwMQXZ4C6wxYD%2BhoqHSEicIB2G0KBIhwDkp7EaR5FSFR4Gflo8gAL5zEiCp9o0yEcB%2BeFEGO7gqOonEcDx2jsHx7CKKo0DOAKSFFKBlr-n%2BYFzKBni3AA%2BjY%2B66fpRR-npf4iBIVR2WQ7C4VAR7qTErDvBYAC02QBCInCOU%2BkDWf%2BQVzKFtm5LOBnSXgP54MMiBKBZUVkAZjGESxmBkdk7EXJx3E8UAA&tokens=false) 

Is this a me problem or the issue still persists with the `no-unused-vars` rule?

The basic snippet lives in the repo linked, in this file `./lib/test.ts` (screenshot below), the original code (and more complex) that was causing me problems lives in `./lib/OpenGraph.ts`

<img width="707" alt="image" src="https://github.com/typescript-eslint/typescript-eslint/assets/364003/60f9f3c1-a26d-4b4b-ad41-32a0a80c0fc0">

### Reproduction Repository Link

https://github.com/Dannymx/shockinglemon.com/tree/bug/no-unused-vars

### Repro Steps

1) clone the repo
2) `pnpm install`
3) `pnpm lint`
4) open `./lib/test.ts` in vscode (this is what I use and where I see the error too)
5) Alternatively also open `./lib/OpenGraph.ts` for a real code example where I'm seeing the error too 

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.2.0` |
| `@typescript-eslint/parser`            | `6.2.0` |
| `@typescript-eslint/scope-manager`     | `X.Y.Z` |
| `@typescript-eslint/typescript-estree` | `X.Y.Z` |
| `@typescript-eslint/type-utils`        | `X.Y.Z` |
| `@typescript-eslint/utils`             | `X.Y.Z` |
| `TypeScript`                           | `5.1.6` |
| `ESLint`                               | `8.45.0` |
| `node`                                 | `v18.17.0` |

**Comment by JoshuaKGoldberg** [comment]
The error doesn't reproduce in the playground: https://typescript-eslint.io/play/#ts=5.1.6&fileType=.tsx&code=C4TwDgpgBAQghgZwJYGMDyAjAVhFwoC8UA3gFBRTAQLABcUNATkgHYDmA3KQL5ekQAPMAHtG%2BUJFiJUAFXDQiAJVyiAJgB5yDYM3YAaLQApiUFnAC2EPQwA2AVzZRu9E2cv0mrTrYcedXpwBKQgA%2BKWR0bFxgUhCuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false. Which means your code is doing something funky.

Furthermore, running `pnpm lint` or `npx eslint lib/test.ts` locally, we get this:

```plaintext
$ npx eslint lib/test.ts
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <5.1.0

YOUR TYPESCRIPT VERSION: 5.1.6

Please only submit bug reports when using the officially supported version.

=============

/Users/josh/repos/shockinglemon.com/lib/test.ts
  7:6   error  'name' is defined but never used  @typescript-eslint/no-unused-vars
  7:12  error  'slug' is defined but never used  @typescript-eslint/no-unused-vars
```

If you check the `pnpm-lock.yaml` file, you'll see that `@typescript-eslint/typescript-estree@5.59.11` shows up.

Then, `npm ls @typescript-eslint/typescript-estree shows (among other output lines):

```plaintext
├─┬ eslint-config-next@13.4.12 -> ./node_modules/.pnpm/eslint-config-next@13.4.12_eslint@8.45.0_typescript@5.1.6/node_modules/eslint-config-next
│ └─┬ @typescript-eslint/parser@5.59.11 -> ./node_modules/.pnpm/@typescript-eslint+parser@5.59.11_eslint@8.45.0_typescript@5.1.6/node_modules/@typescript-eslint/parser
```

`eslint-config-next` will be fixed in https://github.com/vercel/next.js/pull/52848. So this is external. Fun!

For now, you can use https://pnpm.io/package_json#pnpmoverrides.

**Comment by Dannymx** [comment]
For anybody else facing the same issue, here's what I added to `package.json`:
```
...
"pnpm": {
    "overrides": {
      "@typescript-eslint/typescript-estree": "^6.2.0",
      "@typescript-eslint/parser": "^6.2.0"
    }
  },
...
```
